inc/notification.class.php

```diff
@@ -432,6 +433,94 @@ function rawSearchOptions() {
       return $tab;
    }
 
+   /**
+    * Get the massive actions for this object
+    *
+    * @param object|null $checkitem
+    * @return array list of actions
+    */
+   function getSpecificMassiveActions($checkitem = null) {
+
+      $actions = parent::getSpecificMassiveActions($checkitem);
+
+      $actions[__CLASS__.MassiveAction::CLASS_ACTION_SEPARATOR.'add_template'] = _x('button', 'Add notification template');
+      $actions[__CLASS__.MassiveAction::CLASS_ACTION_SEPARATOR.'remove_all_template'] = _x('button', 'Remove all notification template');
+
+      return $actions;
+   }
+
+   /**
+   * @since version 0.85
+   *
+   * @see CommonDBTM::showMassiveActionsSubForm()
+   **/
+   static function showMassiveActionsSubForm(MassiveAction $ma) {
+      switch ($ma->getAction()) {
+         case 'add_template':
+            $notification_notificationtemplate = new Notification_NotificationTemplate();
+            $notification_notificationtemplate->showFormMassiveAction($ma);
+            return true;
+         case 'add_template':
+               return true;
+      }
+      return false;
+   }
+
+
+   static function processMassiveActionsForOneItemtype(MassiveAction $ma, CommonDBTM $item, array $ids) {
+
+      switch ($ma->getAction()) {
+         case 'add_template' :
+            foreach ($ids as $id) {
+               //load notification
+               $notification = new Notification();
+               $notification->getFromDB($id);
+
+               //check if selected template
+               $notification_template = new NotificationTemplate();
+               $notification_template->getFromDB($ma->POST['notificationtemplates_id']);
+
+               if ($notification_template->fields['itemtype'] == $notification->fields['itemtype']) {
+
+                  //check if already exist
+                  $notification_notificationtemplate = new Notification_NotificationTemplate();
+                  $data = [
+                     'mode'                     => $ma->POST['mode'],
+                     'notificationtemplates_id' => $ma->POST['notificationtemplates_id'],
+                     'notifications_id'         => $id
+                  ];
+                  if ($notification_notificationtemplate->getFromDBByCrit($data)) {
+                     $ma->itemDone(Notification::getType(), 0, MassiveAction::ACTION_KO);
+                     $ma->addMessage(NotificationTemplate::getType()." ".$notification_template->getErrorMessage(ERROR_ALREADY_DEFINED)." (".$notification->getLink().")");
+
```

Comment:
I think that if template is already defined, then we should skip addition and consider action as OK.
Indeed, even if action does not change anything in this case, the notification will still be associated with the selected template, so result will be identical to expected one.

Reply:
ok, fixed